orttraining/orttraining/models/runner/pipeline.cc

```diff
@@ -0,0 +1,282 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/models/runner/pipeline.h"
+
+#include <iostream>
+#include <vector>
+#include <cstdint>
+#include <cstdlib>
+#include <stdexcept>
+#include <thread>
+
+#include "gsl/gsl"
+#include "core/framework/ml_value.h"
+#include "orttraining/training_ops/cpu/controlflow/event_pool.h"
+
+namespace onnxruntime {
+namespace training {
+namespace pipeline {
+
+
+Slot::Slot() {
+    type = Empty;
+    batch_id = 0;
+}
+
+bool Slot::IsEmpty() const {
+    return type == Empty;
+}
+
+bool Slot::IsForward() const {
+    return type == Forward;
+}
+
+bool Slot::IsBackward() const {
+    return type == Backward;
+}
+
+void Slot::show() const {
+    int wid = waited_events.size() > 0 ? waited_events[0] : -1;
```

Comment:
>wid [](start = 8, length = 3)

Nit: this might be mistaken as a word. maybe waited_id, recorded_id?

Reply:
Sure.